### DISCUSSION TITLE
Autonomous Naming Convention

### DISCUSSION THREAD

**Discussion by jacueblol** [Patribots4738/Crescendo2024#11] [Ideas]
@Patribots4738/students One of the most important and commonly overlooked aspects of autonomous (and scouting as a whole) is labeling the field in a way that highlights important features of the field and is **easy to understand**. We currently have a good rough draft in place and if you have any questions or concerns, please write them in this discussion. If you have any modifications you would like to make, please consult the [drawing board](https://www.tldraw.com/r/Q1JaBmWD8BmvURaNw1Wdz?viewport=-40,-135,2715,1299&page=page:page). It will update live and is collaborative, so if you encounter questions and have opinions feel free to write them down in there. Here is what it looks like as of this post:
[![image](https://github.com/Patribots4738/Crescendo2024/assets/77470805/754253ba-0617-4540-97f6-b099e4059dbf)](https://www.tldraw.com/r/Q1JaBmWD8BmvURaNw1Wdz?viewport=-121,-148,2715,1299&page=page:page)
Here is an example of what a combination of the current naming conventions can convey:
- `S` `C5` `SR` `C4` `SR` `C3` `SL`
### Translation:
> You shoot your preloaded piece into the speaker, `S`. Then, you pick up `C5`. You go to the position `SR` and shoot in the speaker. You repeat that process for `C4`. For `C3`, however, you go to shoot into the speaker from `SL`, since it would be faster

**Comment by tina-hague** [comment]
Are there any other specific starting positions we can start the game in other than being aligned with W1, W2, & W3?

**Reply by GalexY727** [reply]
W1 - 3 currently represent picking up the game pieces, not starting configurations. In #9 I discuss how we might want to handle starting positions, as they aren't even labeled in this convention as of now.

**Comment by jacueblol** [comment]
I think that the separations of the path should be done at the "thinking points" of the robot. This is an important thing to remember when making auto paths, but I think will be very useful when creating paths in the future. I think we should try to use the Paths portion of Pathplanner to make our autos, not the Autos section of Pathplanner. These paths are generated automatically upon generating the trajectories in Choreo, so no work needs to be done to get them there.
![image](https://github.com/Patribots4738/Crescendo2024/assets/77470805/cd1979f2-7536-497a-9b3c-a97ca6f6eeda)
The reason for this is mainly because of strategy. Recently, we have been experimenting with whether we can use vision in auto to determine our strategy during auto. If everything comes together, we would be able to detect whether or not there should have been a note where we wanted to go, and if the note is missing, we know that something is wrong, and we change our strategy mid-auto. Event markers might help with this, as they won't actually stop the path but can prompt decisions to be made such as looking to see if the piece is there with the camera, but I am not sure as of now how we are going to break the path. A race group might work, but if we see a piece I don't know how to easily break the path and make a new path while guarenteeing that it is optimized and calibrated for our swerve, and would actually work first try.

If we separate our paths in Choreo using stop points when we export, it will save the .traj files separately, allowing us to choose the bot's different thought processes during the auto-using Commands.either() to decide as to which path to use. 
`Example:`
We go to `C5` from at the start of auto, and upon getting to the piece we check if the intake has a piece
  if it has it, we go to shoot it and then pick up `C4`,
  if it doesn't, we go straight to `C4` and skip the shot, since we don't have the piece, since it would be faster if we are racing another robot to the pieces.

```java
PathPlannerPath SC5 = PathPlannerPath.fromPathFile("SC5");
PathPlannerPath C5SC4 = PathPlannerPath.fromPathFile("C5SC");
PathPlannerPath C5C4 = PathPlannerPath.fromPathFile("C5C4");

Command pathWithLogic = AutoBuilder.followPath(SC5)
    .andThen(
        Commands.either(
            AutoBuilder.followPath(C5SC4),
            AutoBuilder.followPath(C5C4),
            intake.hasGamePieceTrigger()
        )
    );
```

**Reply by GalexY727** [reply]
Please keep in mind `PathPlannerPath.fromPathFile` needs to be replaced with `PathPlannerPath.fromChoreoTrajectory` if we are doing Choreo paths.

**Comment by PatribotsProgramming** [comment]
![image](https://github.com/Patribots4738/Crescendo2024/assets/97857553/dffc8013-0061-4852-8bc8-4b9dfac30858)
